### PR TITLE
Adding new zero-shot examples

### LINKED ISSUE #32459: Zero-shot finetuning examples
### Feature request

@amyeroberts @qubvel @NielsRogge 

Does HF team intersted in adding zero-shot finetuning example like https://github.com/SangbumChoi/transformers/tree/grounding_examples/examples/pytorch/zero-shot?

If the team merge these two following open PR then
https://github.com/huggingface/transformers/pull/31828
https://github.com/huggingface/transformers/pull/31964

we can finetune the groundingdino and write something like this.
https://blog.roboflow.com/how-to-fine-tune-paligemma/

I will left some model that you can run 
Model : https://huggingface.co/danelcsb/grounding-dino-tiny-finetuned-cppe-5-10k-steps-no-trainer/tree/main

### Motivation

People can use zero-shot model to finetune there own projects.

### Your contribution

I have made the branch already and ready to generate PR. Need some more modification.
Before
Screenshot 2024-08-06 at 6 08 57 PM
After
Screenshot 2024-08-06 at 6 09 05 PM

### PR DESCRIPTION
**Pull request by SangbumChoi**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/32459

Following two PR should be merged before merging this example!
https://github.com/huggingface/transformers/pull/31828
https://github.com/huggingface/transformers/pull/31964

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by SangbumChoi** [comment]
@qubvel Hi pavel! Is there any standard batch size that hugginface suggest? The reason why I am asking is unlike standard detr, GroundingDino cannot handle batch size 8 for normal single GPU.

**Comment by SangbumChoi** [comment]
@qubvel Hi Pavel this PR is ready for review. However, there is one problem with using evaluation process in `Trainer`.

In order to use the `Trainer` from the transformers we have to get `input_ids` for the post_processing pipeline which is from the tokenizer output. However, transformers `TrainingArgument` `include_input_for_metric` is set to get input variable of first index of class variable which is `pixel_values`. This is hard-coded so I cannot get the input_ids for the input. Also cc @muellerzr.

https://github.com/huggingface/transformers/blob/aefd3e2ae1e2ae5ae555028c0409118bae9e12ee/src/transformers/trainer.py#L3865

There is three ways for this
1. having additional argument to set the name
2. change the order of `forward()`
3. change to main_input_name = getattr(self.model, "input_ids", "input_ids")
WDYT?

**Comment by qubvel** [comment]
Hi @SangbumChoi, thanks for working on this!

> Is there any standard batch size that hugginface suggest? (In my experience maximum batch size was 2 with 24GB) The reason why I am asking is unlike standard detr, GroundingDino cannot handle batch size 8 for normal single GPU.
I think it is fine to have batch size 2 if the model can be trained with such a batch size. In README and a comment we can mention how much memory one would require to run the script and also advise using gradient accumulation, lower precision, and multi-GPU setup (if it is supported).

**Comment by qubvel** [comment]
> However, there is one problem with using evaluation process in Trainer.
In order to use the Trainer from the transformers we have to get input_ids for the post_processing pipeline which is from the tokenizer output. However, transformers TrainingArgument include_input_for_metric is set to get input variable of first index of class variable which is pixel_values. This is hard-coded so I cannot get the input_ids for the input. 

We can probably move these lines
```python
main_input_name = getattr(self.model, "main_input_name", "input_ids")
inputs_decode = self._prepare_input(inputs[main_input_name]) if args.include_inputs_for_metrics else None
```
to a separate method of `Trainer`, for example, `_get_inputs(...)`. Then we will be able to override it in a new `ZeroShotTrainer` class in a script to return required inputs.

**Comment by muellerzr** [comment]
agreed re; `get_inputs`

**Comment by SangbumChoi** [comment]
@qubvel Hi requesting for first review. I will state changes from the original object_detection.py

1. Instead of wandb logging graph, I just added table of GPU consumption (since there is no other architecture to compare)
2. Also as you might know calculating mAP from zero-shot is not accurate. Currently I have made workaround which is the logic that if the output text is directly in `label2id` then match otherwise treat as background object as 0. (We can discuss this if it seems inappropriate). See `def convert_zero_shot_to_coco_format`
3. Added ZeroShotTrainer to override Trainer with `def _get_input_by_name`

Also kudos to @EduardoPach for `backbone_freeze` and `text_backbone_freeze`. Which make the output more stable :)

Before finetune
Screenshot 2024-08-09 at 2 15 04 PM
After finetune
Screenshot 2024-08-09 at 2 15 15 PM

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -4452,7 +4456,7 @@ def prediction_loop(
 
         for step, inputs in enumerate(dataloader):
             loss, logits, labels = self.prediction_step(model, inputs, prediction_loss_only, ignore_keys=ignore_keys)
-            main_input_name = getattr(self.model, "main_input_name", "input_ids")
+            main_input_name = self._get_input_by_name()
             inputs_decode = self._prepare_input(inputs[main_input_name]) if args.include_inputs_for_metrics else None
```
Comment (qubvel) [inline src/transformers/trainer.py:4460]: Lets have both lines moved to a separate method to be able to select even multiple input keys

```python
def _select_inputs_for_validation(inputs):
    main_input_name = getattr(self.model, "main_input_name", "input_ids")
    return inputs[main_input_name]

...
   selected_inputs = _select_inputs_for_validation(inputs)
   inputs_decode = self._prepare_input(selected_inputs) if args.include_inputs_for_metrics else None
```

In ZeroShotTrainer

```python
class ZeroShotTrainer(Trainer):
    def _select_inputs_for_validation(inputs):
        return inputs["input_ids"]
```

In another trainer
```python
class CustomTrainer(Trainer):
    def _select_inputs_for_validation(inputs):
        return {"input_ids": inputs["input_ids"], "pixel_values": inputs["pixel_values"]}
```

**Review by qubvel — COMMENTED** [review]
Hi @SangbumChoi, thanks for iterating! Great work!

I left a comment regarding Trainer, also don't know if it deserves a separate PR. Overall looks very good to me, but I didn't do an in-depth view yet (let's wait for loss PR to be merged). For these examples the tests have to be added and information in README (see files in this [directory](https://github.com/huggingface/transformers/tree/main/examples/pytorch)). Also, feel free to add resources to docs of Grounding DINO. Btw, did you try to fine-tune OwlV2?

**Comment by SangbumChoi** [comment]
@qubvel Sounds good. We can wait until the loss PR get merged.
> Btw, did you try to fine-tune OwlV2?
I didn't aware that we can fine-tune OwlV2, let me see while waiting loss PR!

File: examples/pytorch/zero-shot/run_zero_shot_object_detection.py [hunk 2]
```diff
@@ -0,0 +1,637 @@
...
+
+    # Collect predictions in the required format for metric computation,
+    # model produce boxes in YOLO format, then processor convert them to Pascal VOC format
+    for batch, target_sizes, input_ids in zip(predictions, image_sizes, inputs):
+        batch_logits, batch_boxes = batch[1], batch[2]
+        output = ModelOutput(logits=torch.tensor(batch_logits), pred_boxes=torch.tensor(batch_boxes))
+        post_processed_output = processor.post_process_grounded_object_detection(
+            output, input_ids, box_threshold=box_threshold, text_threshold=text_threshold, target_sizes=target_sizes
+        )
+        post_processed_output = convert_zero_shot_to_coco_format(post_processed_output, label2id)
```
Comment (EduardoPach) [inline examples/pytorch/zero-shot/run_zero_shot_object_detection.py:306]: @SangbumChoi Not sure if this will still work if `random_text_promt=True` because the `class_labels` from targets will not necessarily follow `label2id`

File: examples/pytorch/zero-shot/run_zero_shot_object_detection.py [hunk 3]
```diff
@@ -0,0 +1,637 @@
...
+    """
+    # convert center to corners format
+    torch_label = []
+    for prediction in predictions:
+        scores = prediction["scores"]
+        device = scores.device
+        labels = prediction["labels"]
+        for label in labels:
+            if label in label2id:
+                torch_label.append(label)
```
Comment (EduardoPach) [inline examples/pytorch/zero-shot/run_zero_shot_object_detection.py:147]: I believe you meant `torch_label.append(label2id[label])` right?

**Comment by SangbumChoi** [comment]
Hi @qubvel It is now ready to review this new example. For the CI error it says `unauthorized` I'm not sure how to handle this so it would be appreciated to let me know how to fix this.

**Comment by SangbumChoi** [comment]
Screenshot 2024-09-27 at 4 36 25 PM

**Comment by junxnone** [comment]
> Screenshot 2024-09-27 at 4 36 25 PM

Thanks for share the finetune code.

I test with the code https://github.com/SangbumChoi/transformers on branch grounding_examples.
and only get the lower MAP, it is same as the tensorboard logs in your model [here](https://huggingface.co/danelcsb/grounding-dino-tiny-finetuned-cppe-5-10k-steps-no-trainer/tensorboard)

image

Looks like I missed something...

**Comment by SangbumChoi** [comment]
@junxnone Thanks for the interest. This is because due to difference between zero-shot to coco converted AP. I have fixed in my code but updated in here :(

File: examples/pytorch/zero-shot/run_zero_shot_object_detection_no_trainer.py [hunk 4]
```diff
@@ -0,0 +1,895 @@
...
+        labels = prediction["labels"]
+        for label in labels:
+            if label in label2id:
+                torch_label.append(label2id[label])
+            else:
+                # Give background class
+                torch_label.append(0)
+        prediction["labels"] = torch.Tensor(torch_label).to(dtype=torch.int32).to(device)
+
+    return predictions
```
[the remark this answers is not in the document]
Comment (junxnone) [reply examples/pytorch/zero-shot/run_zero_shot_object_detection_no_trainer.py:150]: I also checkout the pr, and got the same result.

File: examples/pytorch/zero-shot/run_zero_shot_object_detection.py [hunk 5]
```diff
@@ -0,0 +1,637 @@
...
+        random_text_prompt=False,
+    )
+    validation_transform_batch = partial(
+        augment_and_transform_batch,
+        transform=validation_transform,
+        processor=processor,
+        id2label=id2label,
+        label2id=label2id,
+        random_text_prompt=True,
+    )
```
Comment (junxnone) [inline examples/pytorch/zero-shot/run_zero_shot_object_detection.py:588]: I have some confusion here. Why with random_text_prompt=True do I get bad results?

Comment (SangbumChoi) [reply examples/pytorch/zero-shot/run_zero_shot_object_detection.py:588]: This will be deprecated. You understand correctly

File: examples/pytorch/zero-shot/run_zero_shot_object_detection_no_trainer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (SangbumChoi) [reply examples/pytorch/zero-shot/run_zero_shot_object_detection_no_trainer.py:150]: I was meaning I have my own code which is not commited here

**Comment by JamesEmi** [comment]
@qubvel @EduardoPach thanks for your work on this so far (and in general for the wider HF ecosystem :) ). Ended up here after digging around for a train script entrypoint to use for [finetuning grounding dino tiny](https://huggingface.co/IDEA-Research/grounding-dino-tiny). Any updates on why this PR is not merged to the main branch? @EduardoPach could you comment on what training script you used to finetune grounding dino tiny to get [this model](https://huggingface.co/EduardoPacheco/grounding-dino-tiny-aquarium-fine-tune)? Thanks in advance!!!

File: examples/pytorch/zero-shot/run_zero_shot_object_detection.py (hunk 5 above)
Comment (JamesEmi) [reply examples/pytorch/zero-shot/run_zero_shot_object_detection.py:588]: @SangbumChoi thanks a ton for your work so far! I am looking to use this code to finetune GroundingDINO for a critical medical imaging project - and from my understanding of the thread here I am noting some issues I should be careful about. Do you have any points on 'things to remember' when using this code? Or if you have a latest clean version would be super useful to us if you can commit those updates. One other option I am considering is finetuning GroundingDINO using [this repo](https://github.com/longzw1997/Open-GroundingDino); but would still prefer finetuning the model off `transformers`. Thanks!!

**Comment by SangbumChoi** [comment]
@yonigozlan @molbap anybody interested in this PR?

**Comment by SangbumChoi** [comment]
https://github.com/SangbumChoi/transformers/pull/8#issuecomment-4644749929
